File: rs/src/parsers/shape.rs
use nom::number::streaming::{le_u16 as parse_le_u16, le_u8 as parse_u8};
use nom::{IResult as NomResult, Needed};
use std::convert::TryFrom;
use swf_tree as ast;

use crate::parsers::basic_data_types::{
  do_parse_i32_bits, do_parse_u16_bits, do_parse_u32_bits, parse_bool_bits, parse_i32_bits, parse_le_fixed8_p8,
  parse_matrix, parse_s_rgb8, parse_straight_s_rgba8, parse_u16_bits,
};
use crate::parsers::gradient::parse_gradient;

#[derive(PartialEq, Eq, Clone, Copy, Ord, PartialOrd)]
pub enum ShapeVersion {
  Shape1,
  Shape2,
  Shape3,
  Shape4,
}

pub fn parse_glyph(input: &[u8]) -> NomResult<&[u8], ast::Glyph> {
  use nom::bits::bits;
  bits(parse_glyph_bits)(input)
}

pub fn parse_glyph_bits(input: (&[u8], usize)) -> NomResult<(&[u8], usize), ast::Glyph> {
  use nom::combinator::map;

  let (input, fill_bits) = map(do_parse_u32_bits(4), |x| usize::try_from(x).unwrap())(input)?;
  let (input, line_bits) = map(do_parse_u32_bits(4), |x| usize::try_from(x).unwrap())(input)?;
  let (input, records) = parse_shape_record_string_bits(input, fill_bits, line_bits, ShapeVersion::Shape1)?;

  Ok((input, ast::Glyph { records }))
}

pub fn parse_shape(input: &[u8], version: ShapeVersion) -> NomResult<&[u8], ast::Shape> {
  use nom::bits::bits;
  bits(|i| parse_shape_bits(i, version))(input)
}

pub fn parse_shape_bits(input: (&[u8], usize), version: ShapeVersion) -> NomResult<(&[u8], usize), ast::Shape> {
  let (input, styles) = parse_shape_styles_bits(input, version)?;
  let (input, records) = parse_shape_record_string_bits(input, styles.fill_bits, styles.line_bits, version)?;

  Ok((
    input,
    ast::Shape {
      initial_styles: ast::ShapeStyles {
        fill: styles.fill,
        line: styles.line,
      },
      records: records,
    },
  ))
}

// TODO: Rename to InternalShapeStyles or ParserShapeStyles
pub struct ShapeStyles {
  pub fill: Vec<ast::FillStyle>,
  pub line: Vec<ast::LineStyle>,
  pub fill_bits: usize,
  pub line_bits: usize,
}

pub fn parse_shape_styles_bits(input: (&[u8], usize), version: ShapeVersion) -> NomResult<(&[u8], usize), ShapeStyles> {
  use nom::bits::bytes;
  use nom::combinator::map;

  let (input, fill) = bytes(|i| parse_fill_style_list(i, version))(input)?;
  let (input, line) = bytes(|i| parse_line_style_list(i, version))(input)?;
  let (input, fill_bits) = map(do_parse_u32_bits(4), |x| usize::try_from(x).unwrap())(input)?;
  let (input, line_bits) = map(do_parse_u32_bits(4), |x| usize::try_from(x).unwrap())(input)?;

  Ok((
    input,
    ShapeStyles {
      fill,
      line,
      fill_bits,
      line_bits,
    },
  ))
}

pub fn parse_shape_record_string_bits(
  input: (&[u8], usize),
  mut fill_bits: usize,
  mut line_bits: usize,
  version: ShapeVersion,
) -> NomResult<(&[u8], usize), Vec<ast::ShapeRecord>> {
  let mut result: Vec<ast::ShapeRecord> = Vec::new();
  let mut current_input = input;

  loop {
    match parse_u16_bits(current_input, 6) {
      Ok((next_input, record_head)) => {
        if record_head == 0 {
          current_input = next_input;
          break;
        }
      }
      Err(::nom::Err::Incomplete(_)) => return Err(::nom::Err::Incomplete(Needed::Unknown)),
      Err(e) => return Err(e),
    };

    let is_edge = match parse_bool_bits(current_input) {
      Ok((next_input, is_edge)) => {
        current_input = next_input;
        is_edge
      }
      Err(::nom::Err::Incomplete(_)) => return Err(::nom::Err::Incomplete(Needed::Unknown)),
      Err(e) => return Err(e),
    };

    if is_edge {
      let is_straight_edge = match parse_bool_bits(current_input) {
        Ok((next_input, is_straight_edge)) => {
          current_input = next_input;
          is_straight_edge
        }
        Err(::nom::Err::Incomplete(_)) => return Err(::nom::Err::Incomplete(Needed::Unknown)),
        Err(e) => return Err(e),
      };
      let (next_input, edge) = if is_straight_edge {
        parse_straight_edge_bits(current_input)?
      } else {
        parse_curved_edge_bits(current_input)?
      };
      current_input = next_input;
      result.push(ast::ShapeRecord::Edge(edge));
    } else {
      let (next_input, (style_change, style_bits)) =
        parse_style_change_bits(current_input, fill_bits, line_bits, version)?;
      fill_bits = style_bits.0;
      line_bits = style_bits.1;
      result.push(ast::ShapeRecord::StyleChange(style_change));
      current_input = next_input;
    }
  }

  Ok((current_input, result))
}

pub fn parse_curved_edge_bits(input: (&[u8], usize)) -> NomResult<(&[u8], usize), ast::shape_records::Edge> {
  let (input, n_bits) = parse_u16_bits(input, 4).map(|(i, x)| (i, (x as usize) + 2))?;
  let (input, control_x) = parse_i32_bits(input, n_bits)?;
  let (input, control_y) = parse_i32_bits(input, n_bits)?;
  let (input, anchor_x) = parse_i32_bits(input, n_bits)?;
  let (input, anchor_y) = parse_i32_bits(input, n_bits)?;

  Ok((
    input,
    ast::shape_records::Edge {
      delta: ast::Vector2D {
        x: control_x + anchor_x,
        y: control_y + anchor_y,
      },
      control_delta: Some(ast::Vector2D {
        x: control_x,
        y: control_y,
      }),
    },
  ))
}

pub fn parse_straight_edge_bits(input: (&[u8], usize)) -> NomResult<(&[u8], usize), ast::shape_records::Edge> {
  use nom::combinator::{cond, map};

  let (input, n_bits) = map(do_parse_u16_bits(4), |x| (x as usize) + 2)(input)?;
  let (input, is_diagonal) = parse_bool_bits(input)?;
  let (input, is_vertical) = if is_diagonal {
    (input, false)
  } else {
    parse_bool_bits(input)?
  };
  let (input, delta_x) = map(
    cond(is_diagonal || !is_vertical, do_parse_i32_bits(n_bits)),
    Option::unwrap_or_default,
  )(input)?;
  let (input, delta_y) = map(
    cond(is_diagonal || is_vertical, do_parse_i32_bits(n_bits)),
    Option::unwrap_or_default,
  )(input)?;

  Ok((
    input,
    ast::shape_records::Edge {
      delta: ast::Vector2D { x: delta_x, y: delta_y },
      control_delta: None,
    },
  ))
}

pub fn parse_style_change_bits(
  input: (&[u8], usize),
  fill_bits: usize,
  line_bits: usize,
  version: ShapeVersion,
) -> NomResult<(&[u8], usize), (ast::shape_records::StyleChange, (usize, usize))> {
  use nom::combinator::cond;

  let (input, has_new_styles) = parse_bool_bits(input)?;
  let (input, change_line_style) = parse_bool_bits(input)?;
  let (input, change_right_fill) = parse_bool_bits(input)?;
  let (input, change_left_fill) = parse_bool_bits(input)?;
  let (input, has_move_to) = parse_bool_bits(input)?;
  let (input, move_to) = if has_move_to {
    let (input, move_to_bits) = parse_u16_bits(input, 5)?;
    let (input, x) = parse_i32_bits(input, move_to_bits as usize)?;
    let (input, y) = parse_i32_bits(input, move_to_bits as usize)?;
    (input, Some(ast::Vector2D { x, y }))
  } else {
    (input, None)
  };
  let (input, left_fill) = cond(change_left_fill, do_parse_u16_bits(fill_bits))(input)?;
  let (input, right_fill) = cond(change_right_fill, do_parse_u16_bits(fill_bits))(input)?;
  let (input, line_style) = cond(change_line_style, do_parse_u16_bits(line_bits))(input)?;
  let (input, styles) = if has_new_styles {
    let (input, styles) = parse_shape_styles_bits(input, version)?;
    (
      input,
      (
        Some(ast::ShapeStyles {
          fill: styles.fill,
          line: styles.line,
        }),
        styles.fill_bits,
        styles.line_bits,
      ),
    )
  } else {
    (input, (None, fill_bits, line_bits))
  };

  Ok((
    input,
    (
      ast::shape_records::StyleChange {
        move_to: move_to,
        left_fill: left_fill.map(|x| x as usize),
        right_fill: right_fill.map(|x| x as usize),
        line_style: line_style.map(|x| x as usize),
        new_styles: styles.0,
      },
      (styles.1, styles.2),
    ),
  ))
}

pub fn parse_list_length(input: &[u8], allow_extended: bool) -> NomResult<&[u8], usize> {
  let (remaining_input, u8_len) = parse_u8(input)?;
  if u8_len == 0xff && allow_extended {
    parse_le_u16(remaining_input).map(|(i, x)| (i, x as usize))
  } else {
    Ok((remaining_input, u8_len as usize))
  }
}

pub fn parse_fill_style_list(input: &[u8], version: ShapeVersion) -> NomResult<&[u8], Vec<ast::FillStyle>> {
  use nom::multi::count;
  let (input, style_count) = parse_list_length(input, version >= ShapeVersion::Shape2)?;
  count(|i| parse_fill_style(i, version >= ShapeVersion::Shape3), style_count)(input)
}

pub fn parse_fill_style(input: &[u8], with_alpha: bool) -> NomResult<&[u8], ast::FillStyle> {
  use nom::combinator::map;
  let (input, code) = parse_u8(input)?;
  match code {
    0x00 => map(|i| parse_solid_fill(i, with_alpha), ast::FillStyle::Solid)(input),
    0x10 => map(
      |i| parse_linear_gradient_fill(i, with_alpha),
      ast::FillStyle::LinearGradient,
    )(input),
    0x12 => map(
      |i| parse_radial_gradient_fill(i, with_alpha),
      ast::FillStyle::RadialGradient,
    )(input),
    0x13 => map(
      |i| parse_focal_gradient_fill(i, with_alpha),
      ast::FillStyle::FocalGradient,
    )(input),
    0x40 => map(|i| parse_bitmap_fill(i, true, true), ast::FillStyle::Bitmap)(input),
    0x41 => map(|i| parse_bitmap_fill(i, false, true), ast::FillStyle::Bitmap)(input),
    0x42 => map(|i| parse_bitmap_fill(i, true, false), ast::FillStyle::Bitmap)(input),
    0x43 => map(|i| parse_bitmap_fill(i, false, false), ast::FillStyle::Bitmap)(input),
    _ => return Err(nom::Err::Error((input, nom::error::ErrorKind::Switch))),
  }
}

pub fn parse_bitmap_fill(input: &[u8], repeating: bool, smoothed: bool) -> NomResult<&[u8], ast::fill_styles::Bitmap> {
  let (input, bitmap_id) = parse_le_u16(input)?;
  let (input, matrix) = parse_matrix(input)?;
  Ok((
    input,
    ast::fill_styles::Bitmap {
      bitmap_id,
      matrix,
      repeating,
      smoothed,
    },
  ))
}

pub fn parse_focal_gradient_fill(input: &[u8], with_alpha: bool) -> NomResult<&[u8], ast::fill_styles::FocalGradient> {
  let (input, matrix) = parse_matrix(input)?;
  let (input, gradient) = parse_gradient(input, with_alpha)?;
  let (input, focal_point) = parse_le_fixed8_p8(input)?;

  Ok((
    input,
    ast::fill_styles::FocalGradient {
      matrix,
      gradient,
      focal_point,
    },
  ))
}

pub fn parse_linear_gradient_fill(
  input: &[u8],
  with_alpha: bool,
) -> NomResult<&[u8], ast::fill_styles::LinearGradient> {
  let (input, matrix) = parse_matrix(input)?;
  let (input, gradient) = parse_gradient(input, with_alpha)?;

  Ok((input, ast::fill_styles::LinearGradient { matrix, gradient }))
}

pub fn parse_radial_gradient_fill(
  input: &[u8],
  with_alpha: bool,
) -> NomResult<&[u8], ast::fill_styles::RadialGradient> {
  let (input, matrix) = parse_matrix(input)?;
  let (input, gradient) = parse_gradient(input, with_alpha)?;

  Ok((input, ast::fill_styles::RadialGradient { matrix, gradient }))
}

pub fn parse_solid_fill(input: &[u8], with_alpha: bool) -> NomResult<&[u8], ast::fill_styles::Solid> {
  use nom::combinator::map;
  let (input, color) = if with_alpha {
    parse_straight_s_rgba8(input)?
  } else {
    map(parse_s_rgb8, |c| ast::StraightSRgba8 {
      r: c.r,
      g: c.g,
      b: c.b,
      a: 255,
    })(input)?
  };
  Ok((input, ast::fill_styles::Solid { color }))
}

pub fn parse_line_style_list(input: &[u8], version: ShapeVersion) -> NomResult<&[u8], Vec<ast::LineStyle>> {
  use nom::multi::count;
  let (input, style_count) = parse_list_length(input, version >= ShapeVersion::Shape2)?;

  if version >= ShapeVersion::Shape4 {
    count(parse_line_style2, style_count)(input)
  } else {
    count(|i| parse_line_style(i, version >= ShapeVersion::Shape3), style_count)(input)
  }
}

pub fn parse_line_style(input: &[u8], with_alpha: bool) -> NomResult<&[u8], ast::LineStyle> {
  use nom::combinator::map;
  let (input, width) = parse_le_u16(input)?;
  let (input, color) = if with_alpha {
    parse_straight_s_rgba8(input)?
  } else {
    map(parse_s_rgb8, |c| ast::StraightSRgba8 {
      r: c.r,
      g: c.g,
      b: c.b,
      a: 255,
    })(input)?
  };
  Ok((
    input,
    ast::LineStyle {
      width,
      start_cap: ast::CapStyle::Round,
      end_cap: ast::CapStyle::Round,
      join: ast::JoinStyle::Round,
      no_h_scale: false,
      no_v_scale: false,
      no_close: false,
      pixel_hinting: false,
      fill: ast::FillStyle::Solid(ast::fill_styles::Solid { color }),
    },
  ))
}

pub fn parse_line_style2(input: &[u8]) -> NomResult<&[u8], ast::LineStyle> {
  use nom::combinator::map;

  fn cap_style_from_id(cap_style_id: u16) -> ast::CapStyle {
    match cap_style_id {
      0 => ast::CapStyle::Round,
      1 => ast::CapStyle::None,
      2 => ast::CapStyle::Square,
      _ => panic!("UnexpectedCapStyleId: {}", cap_style_id),
    }
  }

  let (input, width) = parse_le_u16(input)?;

  let (input, flags) = parse_le_u16(input)?;
  let pixel_hinting = (flags & (1 << 0)) != 0;
  let no_v_scale = (flags & (1 << 1)) != 0;
  let no_h_scale = (flags & (1 << 2)) != 0;
  let has_fill = (flags & (1 << 3)) != 0;
  let join_style_id = (flags >> 4) & 0b11;
  let start_cap_style_id = (flags >> 6) & 0b11;
  let end_cap_style_id = (flags >> 8) & 0b11;
  let no_close = (flags & (1 << 10)) != 0;
  // (Skip bits [11, 15])

  let start_cap = cap_style_from_id(start_cap_style_id);
  let end_cap = cap_style_from_id(end_cap_style_id);

  let (input, join) = match join_style_id {
    0 => (input, ast::JoinStyle::Round),
    1 => (input, ast::JoinStyle::Bevel),
    2 => {
      let (input, limit) = parse_le_u16(input)?;
      (input, ast::JoinStyle::Miter(ast::join_styles::Miter { limit }))
    }
    _ => panic!("UnexpectedJoinStyleId: {}", join_style_id),
  };

  let (input, fill) = if has_fill {
    parse_fill_style(input, true)?
  } else {
    map(parse_straight_s_rgba8, |color| {
      ast::FillStyle::Solid(ast::fill_styles::Solid { color })
    })(input)?
  };

  Ok((
    input,
    ast::LineStyle {
      width: width,
      fill,
      pixel_hinting,
      no_v_scale,
      no_h_scale,
      no_close,
      join,
      start_cap,
      end_cap,
    },
  ))
}

File: ts/src/lib/parsers/shape.ts
import { ReadableBitStream, ReadableByteStream } from "@open-flash/stream";
import { Incident } from "incident";
import { Sint32, Uint16, Uint2, Uint5, Uint8, UintSize } from "semantic-types";
import {
  CapStyle,
  FillStyle,
  fillStyles,
  FillStyleType,
  Glyph,
  Gradient,
  JoinStyleType,
  LineStyle,
  Matrix,
  Sfixed8P8,
  Shape,
  ShapeRecord,
  shapeRecords,
  ShapeRecordType,
  StraightSRgba8,
  Vector2D,
} from "swf-types";
import { JoinStyle } from "swf-types/join-style";
import { ShapeStyles } from "swf-types/shape-styles";
import { parseMatrix, parseSRgb8, parseStraightSRgba8 } from "./basic-data-types";
import { parseGradient } from "./gradient";

export enum ShapeVersion {
  Shape1 = 1,
  Shape2 = 2,
  Shape3 = 3,
  Shape4 = 4,
}

export function parseGlyph(byteStream: ReadableByteStream): Glyph {
  const bitStream: ReadableBitStream = byteStream.asBitStream();
  const result: Glyph = parseGlyphBits(bitStream);
  bitStream.align();
  return result;
}

export function parseGlyphBits(bitStream: ReadableBitStream): Glyph {
  const fillBits: UintSize = bitStream.readUint32Bits(4);
  const lineBits: UintSize = bitStream.readUint32Bits(4);
  // TODO: Check which shape version to use
  const records: ShapeRecord[] = parseShapeRecordStringBits(bitStream, fillBits, lineBits, ShapeVersion.Shape1);
  return {records};
}

export function parseShape(byteStream: ReadableByteStream, shapeVersion: ShapeVersion): Shape {
  const bitStream: ReadableBitStream = byteStream.asBitStream();
  const result: Shape = parseShapeBits(bitStream, shapeVersion);
  bitStream.align();
  return result;
}

export function parseShapeBits(bitStream: ReadableBitStream, shapeVersion: ShapeVersion): Shape {
  const styles: ParserShapeStyles = parseShapeStylesBits(bitStream, shapeVersion);
  const records: ShapeRecord[] = parseShapeRecordStringBits(
    bitStream,
    styles.fillBits,
    styles.lineBits,
    shapeVersion,
  );
  return {
    initialStyles: {fill: styles.fill, line: styles.line},
    records,
  };
}

export interface ParserShapeStyles {
  fill: FillStyle[];
  line: LineStyle[];
  fillBits: UintSize;
  lineBits: UintSize;
}

export function parseShapeStylesBits(bitStream: ReadableBitStream, shapeVersion: ShapeVersion): ParserShapeStyles {
  const byteStream: ReadableByteStream = bitStream.asByteStream();
  const fill: FillStyle[] = parseFillStyleList(byteStream, shapeVersion);
  const line: LineStyle[] = parseLineStyleList(byteStream, shapeVersion);
  bitStream = byteStream.asBitStream();
  const fillBits: UintSize = bitStream.readUint32Bits(4);
  const lineBits: UintSize = bitStream.readUint32Bits(4);
  return {fill, line, fillBits, lineBits};
}

export function parseShapeRecordStringBits(
  bitStream: ReadableBitStream,
  fillBits: UintSize,
  lineBits: UintSize,
  shapeVersion: ShapeVersion,
): ShapeRecord[] {
  const result: ShapeRecord[] = [];

  while (true) {
    const bytePos: number = bitStream.bytePos;
    const bitPos: number = bitStream.bitPos;
    const head: number = bitStream.readUint16Bits(6);
    if (head === 0) {
      break;
    } else {
      bitStream.bytePos = bytePos;
      bitStream.bitPos = bitPos;
    }

    const isEdge: boolean = bitStream.readBoolBits();
    if (isEdge) {
      const isStraightEdge: boolean = bitStream.readBoolBits();
      if (isStraightEdge) {
        result.push(parseStraightEdgeBits(bitStream));
      } else {
        result.push(parseCurvedEdgeBits(bitStream));
      }
    } else {
      let styles: shapeRecords.StyleChange;
      [styles, [fillBits, lineBits]] = parseStyleChangeBits(bitStream, fillBits, lineBits, shapeVersion);
      result.push(styles);
    }
  }

  return result;
}

export function parseCurvedEdgeBits(bitStream: ReadableBitStream): shapeRecords.Edge {
  const nBits: UintSize = bitStream.readUint16Bits(4) + 2;
  const controlX: Sint32 = bitStream.readSint32Bits(nBits);
  const controlY: Sint32 = bitStream.readSint32Bits(nBits);
  const anchorX: Sint32 = bitStream.readSint32Bits(nBits);
  const anchorY: Sint32 = bitStream.readSint32Bits(nBits);
  return {
    type: ShapeRecordType.Edge,
    controlDelta: {x: controlX, y: controlY},
    delta: {x: controlX + anchorX, y: controlY + anchorY},
  };
}

export function parseStraightEdgeBits(bitStream: ReadableBitStream): shapeRecords.Edge {
  const nBits: UintSize = bitStream.readUint16Bits(4) + 2;
  const isDiagonal: boolean = bitStream.readBoolBits();
  const isVertical: boolean = !isDiagonal && bitStream.readBoolBits();
  const deltaX: Sint32 = isDiagonal || !isVertical ? bitStream.readSint32Bits(nBits) : 0;
  const deltaY: Sint32 = isDiagonal || isVertical ? bitStream.readSint32Bits(nBits) : 0;
  return {
    type: ShapeRecordType.Edge,
    delta: {x: deltaX, y: deltaY},
  };
}

export function parseStyleChangeBits(
  bitStream: ReadableBitStream,
  fillBits: UintSize,
  lineBits: UintSize,
  shapeVersion: ShapeVersion,
): [shapeRecords.StyleChange, [UintSize, UintSize]] {
  const flags: Uint5 = bitStream.readUint32Bits(5);
  const hasMoveTo: boolean = ((flags & (1 << 0)) !== 0);
  const changeLeftFill: boolean = ((flags & (1 << 1)) !== 0);
  const changeRightFill: boolean = ((flags & (1 << 2)) !== 0);
  const changeLineStyle: boolean = ((flags & (1 << 3)) !== 0);
  const hasNewStyles: boolean = ((flags & (1 << 4)) !== 0);

  let moveTo: Vector2D | undefined = undefined;
  if (hasMoveTo) {
    const nBits: UintSize = bitStream.readUint16Bits(5);
    const x: Sint32 = bitStream.readSint32Bits(nBits);
    const y: Sint32 = bitStream.readSint32Bits(nBits);
    moveTo = {x, y};
  }
  const leftFill: UintSize | undefined = changeLeftFill ? bitStream.readUint16Bits(fillBits) : undefined;
  const rightFill: UintSize | undefined = changeRightFill ? bitStream.readUint16Bits(fillBits) : undefined;
  const lineStyle: UintSize | undefined = changeLineStyle ? bitStream.readUint16Bits(lineBits) : undefined;

  let newStyles: ShapeStyles | undefined = undefined;
  if (hasNewStyles) {
    // TODO: Shumway forces `hasNewStyle` to `false` if shapeVersion is `Shape1`, should we do it too?
    // https://github.com/mozilla/shumway/blob/16451d8836fa85f4b16eeda8b4bda2fa9e2b22b0/src/swf/parser/module.ts#L851
    const styles: ParserShapeStyles = parseShapeStylesBits(bitStream, shapeVersion);
    newStyles = {
      fill: styles.fill,
      line: styles.line,
    };
    fillBits = styles.fillBits;
    lineBits = styles.lineBits;
  }

  const styleChangeRecord: shapeRecords.StyleChange = {
    type: ShapeRecordType.StyleChange,
    moveTo,
    leftFill,
    rightFill,
    lineStyle,
    newStyles,
  };

  return [styleChangeRecord, [fillBits, lineBits]];
}

/**
 * Parse a fill style list length or line style list length.
 *
 * @param byteStream Stream to use to parse this list length. Will mutate its state.
 * @param allowExtended Allow extended size (`> 255`). Here are the recommended values:
 *                      - `true` for `DefineShape2`, `DefineShape3`, `DefineShape4`
 *                      - `false` for `DefineShape`
 * @returns List length
 */
export function parseListLength(byteStream: ReadableByteStream, allowExtended: boolean): UintSize {
  const len: UintSize = byteStream.readUint8();
  if (len === 0xff && allowExtended) {
    return byteStream.readUint16LE();
  } else {
    return len;
  }
}

export function parseFillStyleList(
  byteStream: ReadableByteStream,
  shapeVersion: ShapeVersion,
): FillStyle[] {
  const result: FillStyle[] = [];
  const len: UintSize = parseListLength(byteStream, shapeVersion >= ShapeVersion.Shape2);
  for (let i: UintSize = 0; i < len; i++) {
    result.push(parseFillStyle(byteStream, shapeVersion >= ShapeVersion.Shape3));
  }
  return result;
}

export function parseFillStyle(byteStream: ReadableByteStream, withAlpha: boolean): FillStyle {
  const code: Uint8 = byteStream.readUint8();
  switch (code) {
    case 0x00:
      return parseSolidFill(byteStream, withAlpha);
    case 0x10:
      return parseLinearGradientFill(byteStream, withAlpha);
    case 0x12:
      return parseRadialGradientFill(byteStream, withAlpha);
    case 0x13:
      // TODO: Check if this requires shapeVersion >= Shape4
      return parseFocalGradientFill(byteStream, withAlpha);
    case 0x40:
      return parseBitmapFill(byteStream, true, true);
    case 0x41:
      return parseBitmapFill(byteStream, false, true);
    case 0x42:
      return parseBitmapFill(byteStream, true, false);
    case 0x43:
      return parseBitmapFill(byteStream, false, false);
    default:
      throw new Error(`Unexpected fill style code: ${code}`);
  }
}

export function parseBitmapFill(
  byteStream: ReadableByteStream,
  repeating: boolean,
  smoothed: boolean,
): fillStyles.Bitmap {
  const bitmapId: Uint16 = byteStream.readUint16LE();
  const matrix: Matrix = parseMatrix(byteStream);
  return {
    type: FillStyleType.Bitmap,
    bitmapId,
    matrix,
    repeating,
    smoothed,
  };
}

export function parseFocalGradientFill(byteStream: ReadableByteStream, withAlpha: boolean): fillStyles.FocalGradient {
  const matrix: Matrix = parseMatrix(byteStream);
  const gradient: Gradient = parseGradient(byteStream, withAlpha);
  const focalPoint: Sfixed8P8 = Sfixed8P8.fromEpsilons(byteStream.readSint16LE());
  return {
    type: FillStyleType.FocalGradient,
    matrix,
    gradient,
    focalPoint,
  };
}

export function parseLinearGradientFill(byteStream: ReadableByteStream, withAlpha: boolean): fillStyles.LinearGradient {
  const matrix: Matrix = parseMatrix(byteStream);
  const gradient: Gradient = parseGradient(byteStream, withAlpha);
  return {
    type: FillStyleType.LinearGradient,
    matrix,
    gradient,
  };
}

export function parseRadialGradientFill(byteStream: ReadableByteStream, withAlpha: boolean): fillStyles.RadialGradient {
  const matrix: Matrix = parseMatrix(byteStream);
  const gradient: Gradient = parseGradient(byteStream, withAlpha);
  return {
    type: FillStyleType.RadialGradient,
    matrix,
    gradient,
  };
}

export function parseSolidFill(byteStream: ReadableByteStream, withAlpha: boolean): fillStyles.Solid {
  let color: StraightSRgba8;
  if (withAlpha) {
    color = parseStraightSRgba8(byteStream);
  } else {
    color = {...parseSRgb8(byteStream), a: 0xff};
  }
  return {
    type: FillStyleType.Solid,
    color,
  };
}

export function parseLineStyleList(
  byteStream: ReadableByteStream,
  shapeVersion: ShapeVersion,
): LineStyle[] {
  const result: LineStyle[] = [];
  const len: UintSize = parseListLength(byteStream, shapeVersion >= ShapeVersion.Shape2);
  for (let i: UintSize = 0; i < len; i++) {
    if (shapeVersion >= ShapeVersion.Shape4) {
      result.push(parseLineStyle2(byteStream));
    } else {
      result.push(parseLineStyle(byteStream, shapeVersion >= ShapeVersion.Shape3));
    }
  }
  return result;
}

export function parseLineStyle(byteStream: ReadableByteStream, withAlpha: boolean): LineStyle {
  const width: Uint16 = byteStream.readUint16LE();
  const color: StraightSRgba8 = withAlpha ? parseStraightSRgba8(byteStream) : {...parseSRgb8(byteStream), a: 255};
  return {
    width,
    startCap: CapStyle.Round,
    endCap: CapStyle.Round,
    join: {type: JoinStyleType.Round},
    noHScale: false,
    noVScale: false,
    noClose: false,
    pixelHinting: false,
    fill: {
      type: FillStyleType.Solid,
      color,
    },
  };
}

export function parseLineStyle2(byteStream: ReadableByteStream): LineStyle {
  const width: Uint16 = byteStream.readUint16LE();

  const flags: Uint16 = byteStream.readUint16LE();
  const pixelHinting: boolean = (flags & (1 << 0)) !== 0;
  const noVScale: boolean = (flags & (1 << 1)) !== 0;
  const noHScale: boolean = (flags & (1 << 2)) !== 0;
  const hasFill: boolean = (flags & (1 << 3)) !== 0;
  const joinStyleId: Uint2 = ((flags >>> 4) & 0b11) as Uint2;
  const startCapStyleId: Uint2 = ((flags >>> 6) & 0b11) as Uint2;
  const endCapStyleId: Uint2 = ((flags >>> 8) & 0b11) as Uint2;
  const noClose: boolean = (flags & (1 << 10)) !== 0;
  // (Skip bits [11, 15])

  let join: JoinStyle;
  switch (joinStyleId) {
    case 0:
      join = {type: JoinStyleType.Round};
      break;
    case 1:
      join = {type: JoinStyleType.Bevel};
      break;
    case 2:
      join = {type: JoinStyleType.Miter, limit: Sfixed8P8.fromEpsilons(byteStream.readSint16LE())};
      break;
    default:
      throw new Incident("UnexpectedJoinStyleId", {id: joinStyleId});
  }

  let fill: FillStyle;
  if (hasFill) {
    fill = parseFillStyle(byteStream, true);
  } else {
    fill = {type: FillStyleType.Solid, color: parseStraightSRgba8(byteStream)};
  }

  return {
    width,
    fill,
    pixelHinting,
    noVScale,
    noHScale,
    noClose,
    join,
    startCap: capStyleFromId(startCapStyleId),
    endCap: capStyleFromId(endCapStyleId),
  };
}

export function capStyleFromId(capStyleId: Uint2): CapStyle {
  switch (capStyleId) {
    case 0:
      return CapStyle.Round;
    case 1:
      return CapStyle.None;
    case 2:
      return CapStyle.Square;
    default:
      throw new Incident("UnexpectedCapStyleId", {id: capStyleId});
  }
}
